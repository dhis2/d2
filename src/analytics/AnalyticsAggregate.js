import AnalyticsBase from './AnalyticsBase';

/**
 * @extends module:analytics.AnalyticsBase
 *
 * @description
 * Analytics aggregate class used to request aggregate analytics data from Web API.
 *
 * @memberof module:analytics
 *
 * @see https://docs.dhis2.org/master/en/developer/html/webapi_analytics.html
 */
class AnalyticsAggregate extends AnalyticsBase {
    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics data value set data from the API.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     * d2.analytics.aggregate.getDataValueSet(req)
     * .then(console.log);
     */
    getDataValueSet(req) {
        return this.get(req.withPath('dataValueSet'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the raw data from the API.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw', 'PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu');
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31')
     *  .withFormat('xml');
     *
     *  d2.analytics.aggregate.getRawData(req)
     *  .then(console.log);
     */
    getRawData(req) {
        return this.get(req.withPath('rawData'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the SQL statement used to query the database.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw', 'PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu');
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.aggregate.getDebugSql(req);
     *  .then(console.log);
     */
    getDebugSql(req) {
        return this.get(req.withPath('debug/sql'));
    }
}

export default AnalyticsAggregate;
