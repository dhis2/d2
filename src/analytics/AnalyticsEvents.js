import AnalyticsBase from './AnalyticsBase';

/**
 * @extends module:analytics.AnalyticsBase
 *
 * @description
 * Analytics events class used to request analytics events data from Web API.
 *
 * @memberof module:analytics
 *
 * @see https://docs.dhis2.org/master/en/developer/html/webapi_event_analytics.html
 */
class AnalyticsEvents extends AnalyticsBase {
    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics aggregate data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.events.getAggregate(req)
     *  .then(console.log);
     */
    getAggregate(req) {
        return this.get(req.withPath('events/aggregate'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics count data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.events.getCount(req)
     *  .then(console.log);
     */
    getCount(req) {
        return this.get(req.withPath('events/count'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     * Must contain clusterSize and bbox parameters.
     *
     * @returns {Promise} Promise that resolves with the analytics cluster data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31')
     *  .withClusterSize(100000)
     *  .withBbox('-13.2682125,7.3721619,-10.4261178,9.904012');
     *
     *  d2.analytics.events.getCluster(req)
     *  .then(console.log);
     */
    getCluster(req) {
        return this.get(req.withPath('events/cluster'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics query data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.events.getQuery(req)
     *  .then(console.log);
     */
    getQuery(req) {
        return this.get(req.withPath('events/query'));
    }
}

export default AnalyticsEvents;
