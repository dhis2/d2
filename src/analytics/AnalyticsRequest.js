import AnalyticsRequestDimensionsMixin from './AnalyticsRequestDimensionsMixin';
import AnalyticsRequestFiltersMixin from './AnalyticsRequestFiltersMixin';
import AnalyticsRequestPropertiesMixin from './AnalyticsRequestPropertiesMixin';
import AnalyticsRequestBase from './AnalyticsRequestBase';

/**
 * @description
 * Class for constructing a request object to use for communicating with the analytics API endpoint.
 *
 * @param {!Object} options Object with analytics request options
 *
 * @memberof module:analytics
 *
 * @extends module:analytics.AnalyticsRequestDimensionsMixin
 * @extends module:analytics.AnalyticsRequestFiltersMixin
 * @extends module:analytics.AnalyticsRequestPropertiesMixin
 * @extends module:analytics.AnalyticsRequestBase
 */
class AnalyticsRequest extends AnalyticsRequestDimensionsMixin(
    AnalyticsRequestFiltersMixin(AnalyticsRequestPropertiesMixin(AnalyticsRequestBase)),
) {
    /**
     * Extracts dimensions and filters from an analytic object model and add them to the request
     *
     * @param {Object} model The analytics object model from which extract the dimensions/filters
     *
     * @returns {AnalyticsRequest} A new instance of the class for chaining purposes
     *
     * @example
     * const req = new d2.analytics.request()
     *    .fromModel(model);
     *
     * // dimension=pe:LAST_12_MONTH&dimension=dx:fbfJHSPpUQD;cYeuwXTCPkU;Jtf34kNZhzP;hfdmMSPBgLG&filter=ou:ImspTQPwCqd
     */
    fromModel(model) {
        let request = this;

        // extract dimensions from model
        const columns = model.columns || [];
        const rows = model.rows || [];

        columns.concat(rows).forEach((d) => {
            let dimension = d.dimension;

            if (d.filter) {
                dimension += `:${d.filter}`;
            }

            request = request.addDimension(dimension, d.items.map(item => item.id));
        });

        // extract filters from model
        const filters = model.filters || [];

        filters.forEach(
            f => (request = request.addFilter(f.dimension, f.items.map(item => item.id))),
        );

        return request;
    }
}

export default AnalyticsRequest;
