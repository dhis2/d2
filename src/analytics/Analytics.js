/**
 * @module analytics
 */
import AnalyticsAggregate from './AnalyticsAggregate'
import AnalyticsEvents from './AnalyticsEvents'
import AnalyticsRequest from './AnalyticsRequest'
import AnalyticsResponse from './AnalyticsResponse'

/**
 * @description
 * Analytics class used to request analytics data from Web API.
 *
 * @requires analytics.AnalyticsAggregate
 * @requires analytics.AnalyticsEvents
 * @requires analytics.AnalyticsRequest
 * @requires analytics.AnalyticsResponse
 *
 * @example
 * const d2Analytics = new Analytics(
 *  new AnalyticsAggregate(),
 *  new AnalyticsEvents(),
 *  AnalyticsRequest,
 *  AnalyticsResponse
 * )
 *
 * @memberof module:analytics
 * @see {@link https://docs.dhis2.org/master/en/developer/html/webapi_analytics.html} Analytics API documentation
 * @see {@link https://docs.dhis2.org/master/en/developer/html/webapi_event_analytics.html} Event analytics API documentation
 */
class Analytics {
    /**
     * @param {!module:analytics.AnalyticsAggregate} analyticsAggregate The AnalyticsAggregate instance
     * @param {!module:analytics.AnalyticsEvents} analyticsEvents The AnalyticsEvents instance
     * @param {!module:analytics.AnalyticsRequest} analyticsRequest The AnalyticsRequest class
     * @param {!module:analytics.AnalyticsResponse} analyticsResponse The AnalyticsResponse class
     */
    constructor(
        analyticsAggregate,
        analyticsEvents,
        analyticsRequest,
        analyticsResponse
    ) {
        this.aggregate = analyticsAggregate
        this.events = analyticsEvents
        this.request = analyticsRequest
        this.response = analyticsResponse
    }

    /**
     * @static
     *
     * @description
     * Get a new instance of the Analytics object. This will function as a singleton, once Analytics object
     * has been created when requesting getAnalytics again the original version will be returned.
     *
     * @returns {Analytics} Object with the analytics interaction properties
     *
     * @example
     * const d2Analytics = d2.getAnalytics();
     */
    static getAnalytics() {
        if (!Analytics.getAnalytics.analytics) {
            Analytics.getAnalytics.analytics = new Analytics(
                new AnalyticsAggregate(),
                new AnalyticsEvents(),
                AnalyticsRequest,
                AnalyticsResponse
            )
        }

        return Analytics.getAnalytics.analytics
    }
}

export default Analytics
