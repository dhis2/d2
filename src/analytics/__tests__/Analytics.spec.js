import Analytics from '../Analytics';
import AnalyticsAggregate from '../AnalyticsAggregate';
import AnalyticsEvents from '../AnalyticsEvents';
import AnalyticsRequest from '../AnalyticsRequest';

describe('Analytics', () => {
    let analytics;

    beforeEach(() => {
        analytics = new Analytics(new AnalyticsAggregate(), new AnalyticsEvents(), AnalyticsRequest);
    });

    it('should create an instance of Analytics', () => {
        expect(Analytics.getAnalytics()).toBeInstanceOf(Analytics);
    });

    it('should not be allowed to be called without new', () => {
        expect(() => Analytics()).toThrowErrorMatchingSnapshot();
    });

    it('should contain an instance of AnalyticsAggregate', () => {
        expect(analytics.aggregate).toBeInstanceOf(AnalyticsAggregate);
    });

    it('should contain an instance of AnalyticsEvents', () => {
        expect(analytics.events).toBeInstanceOf(AnalyticsEvents);
    });

    it('should contain a reference to AnalyticsRequest', () => {
        expect(analytics.request).toBe(AnalyticsRequest);
    });

    describe('getAnalytics', () => {
        it('should return the same instance on consecutive requests', () => {
            expect(Analytics.getAnalytics()).toBe(Analytics.getAnalytics());
        });
    });
});
