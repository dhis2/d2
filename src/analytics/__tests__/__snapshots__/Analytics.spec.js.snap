// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics should not be allowed to be called without new 1`] = `"Class constructor Analytics cannot be invoked without 'new'"`;
